holoviews/core/util.py

```diff
@@ -24,6 +26,57 @@
 except ImportError:
     dd = None
 
+
+
+
+class HashGenerator(json.JSONEncoder):
+    """
+    Extends JSONEncoder to generate a hashable string for as many types
+    of key as possible. These keys are supplied by widgets and Stream
+    parameters and may be nested. If an unrecognized object type is
+    supplied, the objects id is used - this may be an issue when dealing
+    with mutable objects. For this reason, it is important to support
+    common mutable types such as pandas Dataframes and numpy arrays
+    (supported).
+
+    One limitation of this approach is that dictionaries with composite
+    keys (e.g tuples) are not supported due to the JSON spec.
+
+    Used by keyhash to help memoize DynamicMap output.
+    """
+    string_hashable = (dt.datetime,)
+    repr_hashable = ()
+
+    def default(self, obj):
+        if isinstance(obj, set):
+            return hash(frozenset(obj))
+        elif isinstance(obj, np.ndarray):
+            return obj.tolist()
+        if pd and isinstance(obj, (pd.Series, pd.DataFrame)):
+            return sorted(obj.to_dict().items())
+        elif isinstance(obj, self.string_hashable):
+            return str(obj)
+        elif isinstance(obj, self.repr_hashables):
+            return repr(obj)
+        try:
+            return hash(obj)
+        except:
+            return id(obj)
+
+
+
+def keyhash(key, as_string=False):
```

Comment:
> Or is this ``hashable_string``?

No. I'll get rid of the ``as_string`` argument as I only needed it temporarily for debugging. The output is an integer, as returned by the Python ``hash`` function.

I am happy to rename it. In the context it is used (memoisation) the argument is what we have always called a 'key'. I'll just call it ``obj`` here... but then this function then needs a new name as ``hash`` clashes with the built-in. Again, any suggestions?